Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt

```diff
@@ -20,23 +29,64 @@ class RecoveryCertificateRepository @Inject constructor(
     private val dispatcherProvider: DispatcherProvider,
     private val qrCodeExtractor: RecoveryCertificateQRCodeExtractor,
     valueSetsRepository: ValueSetsRepository,
+    private val storage: RecoveryCertificateStorage,
 ) {
 
-    val certificates: Flow<Set<RecoveryCertificateWrapper>> = emptyFlow()
+    private val internalData: HotDataFlow<Set<RecoveryCertificateContainer>> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.IO,
+        sharingBehavior = SharingStarted.Lazily,
+    ) {
+        storage.recoveryCertificates
+            .map { recoveryCertificate ->
+                RecoveryCertificateContainer(
+                    data = recoveryCertificate,
+                    qrCodeExtractor = qrCodeExtractor
+                )
+            }
+            .toSet()
+            .also { Timber.tag(TAG).v("Restored recovery certificate data: %s", it) }
+    }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing test data.") }
+            .onEach { recoveryCertificates ->
+                Timber.tag(TAG).v("Recovery Certificate data changed: %s", recoveryCertificates)
+                storage.recoveryCertificates = recoveryCertificates.map { it.data }.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot recovery certificate data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.IO)
+    }
+
+    val certificates: Flow<Set<RecoveryCertificateWrapper>> = combine(
+        internalData.data,
+        valueSetsRepository.latestVaccinationValueSets // TODO use corrects sets
+    ) { containers, currentValueSet ->
+        containers.map { RecoveryCertificateWrapper(currentValueSet, it) }.toSet()
+    }
 
     suspend fun requestCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("requestCertificate(qrCode=%s)", qrCode)
         throw NotImplementedError()
     }
 
     suspend fun deleteCertificate(identifier: RecoveryCertificateIdentifier) {
-        Timber.tag(TAG).d("deleteTestCertificate(identifier=%s)", identifier)
-        throw NotImplementedError()
+        Timber.tag(TAG).d("deleteCertificate(identifier=%s)", identifier)
+        internalData.updateBlocking {
+            mapNotNull { if (it.certificateId == identifier) null else it }.toSet()
```

Comment:
`certificateId` is different from `identifier`

* `certificateId` is the ID field in the scanned certificate JSON
* `identifier` is a unique ID we come up with

We could make them the same, but we didn't do this for the other certificates as we would have to rely on them being unique (which I don't think we should do, as we have seen, countries do all sorts of different things that are not according to schema specs...).

Reply:
introduced uuid for client side identification